CARDS-1518: Import Cardiac Rehab phase 2 questionnaire changes

acrowthe opened the pull request:
- Fix advanced conditional handling
  - Add support for commas in conditional strings
  - Fix conditions which depend on multiple answers
- Clean some section and question names that were breaking the import

New forms that need to be tested:
- Changes in Health Survey
- End of Study Physical Assessment
- Recent Activities Survey
- WHODAS
Questionnaire definitions can be found in the pinned message in the `#cardiac-rehab-dev` slack channel, `New Questionnaires` sheet.

All the other questionnaires have not been changed

#### Note:

`CARDS-1518` is based on the `stable-0.8.x` branch to keep the changes compatible with the version currently in production. After building, run with:
```
java -jar distribution/target/lfs-*.jar -Dsling.run.modes=cardiac_rehab
```

acrowthe committed:
CARDS-1518: Import Cardiac Rehab phase 2 questionnaire changes
- Fix advanced conditional handling
  - Add support for commas in conditional strings
  - Fix conditions which depend on multiple answers
- Clean some section and question names that were breaking the import

sashaandjic commented:
WHODAS has no visits numbers. There should be visits 9, 10, 12, 14, 16, 18, 21.

sashaandjic commented:
Changes in health questionnaire:
`I prefer not to answer` option should behave same as `None of the above`
![image](https://user-images.githubusercontent.com/23243244/143791756-1b058d70-cb6f-483b-bded-fc8c64eeb20d.png)

sashaandjic commented:
Changes in Health Survey - other options is missing. It should be positioned above `None of the above`
![image](https://user-images.githubusercontent.com/23243244/143791924-f9d9ae5b-7f32-4c2e-9e4b-6932dfdecd31.png)

sashaandjic commented:
`Specify date` for cancer is not shown
![image](https://user-images.githubusercontent.com/23243244/143792350-c369b8e2-a0b7-4228-979a-f6fa4738eee5.png)

sashaandjic commented:
Changes in Health Survey - the spreadsheet has this `In the last 30 days, did you experience a fall?`, however the questinnaire has `In the last 3 months, did you experience a fall?`

acrowthe committed:
CARDS-1518: Import Cardiac Rehab phase 2 questionnaire changes
Minor fixes and improvements
- Add `not applicable` flag to `I prefer not to answer` questions
- Fix visit number and frequencies on new questionnaires
- Include cancer date and fall time changes

acrowthe commented:
> WHODAS and Changes in Health Survey has no visits numbers. There should be visits 9, 10, 12, 14, 16, 18, 21.
Fixed for Changes in Health, WHODAS and Recent Activities

> I prefer not to answer option should behave same as None of the above
"I prefer not to answer" options now act as "Not applicable" (clears every single option, not just options above) in Recent Activities and Changes in Health Surveys

> Other options is missing. It should be positioned above None of the above; as a consequence Specify other diagnosis field is not shown - HOWEVER I LIKE THE WORKAROUND
Done intentionally, "Other" options and follow up conditional questions are replaced by free text options in all Cardiac rehab questionnaires

> Specify date for cancer is not shown
Fixed

> Not sure about Congestive Heart Failure - would it look better as Congestive heart failure?
Unchanged for now.

> Changes in Health Survey - the spreadsheet has this In the last 30 days, did you experience a fall?, however the questinnaire has In the last 3 months, did you experience a fall?
Changed to 30 days

sashaandjic commented:
Visit number question is still missing from **WHODAS** and **Changes in Health Survey** questionnaires

![image](https://user-images.githubusercontent.com/23243244/143809189-3166e2e5-9adb-4316-b27c-ef1b542e25d9.png)

sashaandjic commented:
All other reported problems are fixed.

acrowthe committed:
CARDS-1518: Import Cardiac Rehab phase 2 questionnaire changes
- Add missing Visit number questions to Changes in Health,
  Recent Activities and WHODAS

acrowthe commented:
> Visit number question is still missing from WHODAS and Changes in Health Survey questionnaires

Fixed, added visit number question to WHODAS, Changes in Health and Recent Activities

sashaandjic commented:
QUESTION: @k-pace Should this question have multivalues selected?
![image](https://user-images.githubusercontent.com/23243244/143948527-765f0695-dff2-44b5-ac8c-f7d99f25aea0.png)

sashaandjic commented:
This question should allow multi value selection (max 3).

<html>
<body>
<!--StartFragment--><google-sheets-html-origin><style type="text/css"><!--td {border: 1px solid #ccc;}br {mso-data-placement:same-cell;}--></style>

chs6fitness_# | String (max 3) |   | What are your fitness goals for the next 30 days? |   | Gain weight Lose weight Build muscle mass Tone muscles Improve balance and dexterity Improve endurance Improve health Improve sports performance Maintain current fitness Maintain current shape and weight Maintain current health Recover from an in jury Other I don’t have any fitness goals I prefer not to answer
-- | -- | -- | -- | -- | --


<!--EndFragment-->
</body>
</html>

![image](https://user-images.githubusercontent.com/23243244/143949047-85595b07-af75-45d4-8247-a2b5ea4640d2.png)

acrowthe committed:
CARDS-1518: Import Cardiac Rehab phase 2 questionnaire changes
Add Max Answers support and implement in Changes in Health questionnaire

acrowthe commented:
> This question should allow multi value selection (max 3).

Max Answers column added to the spreadsheet & implemented in the importer. Fixed

sashaandjic reviewed (approved):
tested

acrowthe committed:
CARDS-1531: Update Date Conditionals for Cardiac Rehab phase 2
- Add conditional dates for Changes in Health multi-select questions
- Fix Recent Activities survey multi-select that was incorrectly single select

sashaandjic commented:
**Changes in Health Survey**
1. Question `Were you diagnosed with any of the following in the past 30 days?`
- missing `Specify date of ` + other +  `diagnosis`.
- there is a lone `cancerDate_#` - `Specify date`. It is always present. It should be renamed to `Specify date of cancer diagnosis` and made conditional `chs1diagnosis = Cancer`.
![image](https://user-images.githubusercontent.com/23243244/144174386-f0f6ee9c-f706-4f82-ac52-98a501e33a94.png)
2. Question - `chs1lungConditionDate_#`  - `When did you get diagnosed with another condition in the past 30 days` should be split into 3+ conditional dates: `Specify date of Chronic bronchitis`, `Specify date of Chronic obstructive pulmonary disease (COPD)`, `Specify date of Emphysema`, `Specify date of` + other
3. Question  `chs1clinicalTest_#` - replace "list+input" with "list and add `Other` option. Also, these 2 conditional questions seem to be missing 
![image](https://user-images.githubusercontent.com/23243244/144177633-bff123a8-e193-492b-b431-07e3e0fb8904.png)
4. Question `chs6fitness_#` - `What are your fitness goals for the next 30 days? ` -  replace "list+input" with "list and add `Other` option. Also, `chs6fitnessOther_#` - `Specify other fitness goal` with condition `Y if chs6fitness = Other`  seem to be missing

acrowthe committed:
CARDS-1531: Update Date Conditionals for Cardiac Rehab phase 2
- Convert "Other" fields in the Changes in Health survey to recurrent sections
- Add `recurrent` column support to add a recurrent subsection
  - If a child question is conditional, the recurrent section copies that condition
  - Does not end any parent sections specified by sub-report
  - Ends if the `recurrent` column has the value "end"

acrowthe committed:
CARDS-1531: Update Date Conditionals for Cardiac Rehab phase 2

Cleanup old comment

k-pace commented:
## WHODAS
- [ ] Move questionnaire preamble to after Visit # selection (done in spreadsheet, requires reimport):
<img width="656" alt="image" src="https://user-images.githubusercontent.com/33154316/144477898-f9be869a-496e-45d6-993b-33a3a45a98ee.png">

## Changes in Health
- [ ] **BUG**: Lung Condition selection persists after deselecting "I was diagnosed with a new illness or medical condition" (`chs1`) -> "Lung condition" (`chs1diagnosis`)
![image](https://user-images.githubusercontent.com/33154316/144484641-b6e1f577-dd35-4ad1-ade0-5a6aa776291c.png)
- [ ] **BUG**: `cane_startedDate` (along with other conditional dates nested under `chs5mobility`) persists after deselecting 'Broken bone' or 'Change in lower limb arthritis' in `chs3lifeEvents` or Joint replacement / Other bone, joint or muscle surgery under `chs1operations`
![image](https://user-images.githubusercontent.com/33154316/144485795-6e111a49-3d5f-483d-8e5c-a0d64d37f940.png)

- [ ] **Unexpected Behaviour**: `chs2hospitalVisit` -> 'No' should act like "None of the Above" to clear all inputs. This is also present in `chs5mobility`
![image](https://user-images.githubusercontent.com/33154316/144485026-576400ed-fd88-45a0-96e3-d4cb689b6d9d.png)

sashaandjic commented:
`Respiratory condition ` seems to cancel out conditional date fields for `Lung condition`
https://user-images.githubusercontent.com/23243244/144491839-5ad9862d-21c0-4152-81ec-4a2370d413e7.mp4

k-pace commented:
> `Respiratory condition ` seems to cancel out conditional date fields for `Lung condition` https://user-images.githubusercontent.com/23243244/144491839-5ad9862d-21c0-4152-81ec-4a2370d413e7.mp4

@sashaandjic I don't think it does; the order just becomes strange because of the number of conditionals - keep scrolling down with both selected and you should still see the lung condition conditionals

sashaandjic commented:
> @sashaandjic I don't think it does; the order just becomes strange because of the number of conditionals - keep scrolling down with both selected and you should still see the lung condition conditionals

Ah, I see. The respiratory option actually adds more than 1 conditional field, and I assumed those were not related to the respiratory option.

sashaandjic commented:
Changes in Health Survey:
Can we group together all 4 questions with condition `chs1diagnosis = Respiratory condition (e.g. cold, cough, flu)`?

sashaandjic commented:
It would be much easier if the list is alphabetically sorted. Few options need to be moved to another spot:
-`Pre-diabetes or impaired glucose tolerance` 
- `Respiratory condition (e.g. cold, cough, flu)`
- `Lung condition`
![image](https://user-images.githubusercontent.com/23243244/144501927-a4afda57-c89e-4069-93f9-d6893e24b078.png)

acrowthe committed:
CARDS-1531: Update Date Conditionals for Cardiac Rehab phase 2
- Fix nested conditional questions not hiding when intended
- Re-order some questions and options to improve form readability
- Add the "None of the above" behaviour flag to "no" options

k-pace commented:
> It would be much easier if the list is alphabetically sorted. Only 3 options need to be moved to the proper spot:
> 
> * `Pre-diabetes or impaired glucose tolerance`
> * `Respiratory condition (e.g. cold, cough, flu)`
> * `Lung condition`
>   ![image](https://user-images.githubusercontent.com/23243244/144501927-a4afda57-c89e-4069-93f9-d6893e24b078.png)

I did this in the spreadsheet; not sure if @acrowthe wants to re-run the import for this or not

acrowthe commented:
> I did this in the spreadsheet; not sure if @acrowthe wants to re-run the import for this or not

Still working on the conditionals, so I'll definitely reimport for order

acrowthe committed:
CARDS-1531: Update Date Conditionals for Cardiac Rehab phase 2
- Add all missing complex conditionals to fix nested conditional bug
- Reorder some options to be alphabetical

sashaandjic commented:
Recurring `other` section/questions are not showing up in view mode
![image](https://user-images.githubusercontent.com/23243244/144673110-f2b10a7a-bab1-45f1-9362-b24b32ab7f14.png)
![image](https://user-images.githubusercontent.com/23243244/144673508-468e3a64-fbd4-4d61-bbe2-3e4dd2aeb5f5.png)


The other problem is that I should not be able to select more than 3 options in the above question, however having set other as a recurring section I can create many more.

marta- commented:
> I should not be able to select more than 3 options in the above question, however having set other as a recurring section I can create many more.

I suggest we let this slide for now. The data is being entered by trusted users, and there's no support for limiting the number of recurring sections at the moment.

marta- commented:
> Recurring `other` section/questions are not showing up in view mode

I think what we're seeing here might be a manifestation of [CARDS-1527](https://phenotips.atlassian.net/browse/CARDS-1527): _Subsections are not rendered in a view mode if their parent Section does not have direct Question descendants with existing answers_ .

sashaandjic commented:
The problem seems to be out of scope. I see the same issue in the Baseline Medication form, which has nothing to do with changes in this PR. I am removing the `bug` tag.
![image](https://user-images.githubusercontent.com/23243244/144677502-b2e4d256-3dcd-4dd0-8284-3bd2a43541e1.png)
![image](https://user-images.githubusercontent.com/23243244/144677554-b98f685e-35c2-407b-8f18-84a3b97709d7.png)

marta- commented:
Just noticed: the question in the "Other goal" section shouldhave maxAnswers=1 (only one goal per entry)

sashaandjic commented:
> Just noticed: the question in the "Other goal" section shouldhave maxAnswers=1 (only one goal per entry)
Going with maxAnswers=1 would prevent the problem of entering more than 3 options.

marta- commented:
> > Just noticed: the question in the "Other goal" section shouldhave maxAnswers=1 (only one goal per entry)
> 
> Going with maxAnswers=1 would prevent the problem of entering more than 3 options.

Not quite, you can add as many "Other" sections as you want.

sashaandjic commented:
I am out of ideas on how to make this new form more pretty by moving the text "think back...".
Tried some options and nothing worked.
![image](https://user-images.githubusercontent.com/23243244/144679995-bfe413ca-d201-476e-b35f-c2df3950dab0.png)

marta- commented:
> I am out of ideas on how to make this new form more pretty by moving the text **"Think back..."**. Tried some options and nothing worked.

There's nothing we can do in this old version. in the latest dev we could use an info card for "Think back". In the 1400 branch this section would be a question matrix. But for now we'll just have to leave it as is.

The only thing that would balance the layout is adding some sort of section title just above the "Visit" question.

acrowthe committed:
CARDS-1531: Update Date Conditionals for Cardiac Rehab phase 2
- Fix Section-in-sections causing issues on form view mode
- Fix multiple answer other fields
- Add section title to WHODAS form

marta- commented:
Is compact layout better for WHODAS?

![image](https://user-images.githubusercontent.com/651980/144685450-0cf814b9-c640-486e-89bb-70886ed2bebc.png)

marta- reviewed (approved):
Looks good to me after basic testing

sashaandjic commented:
@marta- I like the compact layout for WHODAS.
On the Baseline `Medications questionnaire` I still do not see recurring sections when in view mode. I guess it is out of scope?

sashaandjic reviewed (approved):
tested
